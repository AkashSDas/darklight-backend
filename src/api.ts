import "./passport";

import cookieParser from "cookie-parser";
import cors from "cors";
import { config } from "dotenv";
import express from "express";
import expressSession from "express-session";
import morgan from "morgan";
import passport from "passport";

// Load env vars
if (process.env.NODE_ENV != "production") config();

/** Express app */
export var app = express();

// =====================================
// Middlewares
// =====================================

app.use(morgan("dev"));
app.use(cors({ origin: process.env.FRONTEND_URL, credentials: true }));
app.use(express.json()); // for parsing incoming data
app.use(express.urlencoded({ extended: true })); // parses incoming requests with urlencoded payloads

app.use(cookieParser()); // Parse Cookie header and populate req.cookies with an object keyed by the cookie names
app.use(
  expressSession({
    secret: process.env.COOKIE_SESSION_SECRET,
    resave: true,
    saveUninitialized: true,
  })
);

app.use(passport.initialize());
app.use(passport.session());

// =====================================
// Routes
// =====================================

app.get("/api/v2/testing/test", function testController(_req, res) {
  res.status(200).json({ message: "Hello World!" });
});

app.use("/api/v2/auth", require("./routes/auth.route").router);

app.all("*", function handleUnknownRoutes(req, res) {
  return res.status(404).json({
    message: `Cannot find ${req.originalUrl} on this server!`,
  });
});
