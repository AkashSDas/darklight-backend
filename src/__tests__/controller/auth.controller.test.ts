import { MongoMemoryServer } from "mongodb-memory-server";
import mongoose from "mongoose";
import supertest from "supertest";

import { afterEach, beforeEach, describe, expect, it } from "@jest/globals";

import { app } from "../../api";
import User from "../../models/user.schema";

describe("auth controller", () => {
  beforeEach(async function connectToMongoDB() {
    let mongoServer = await MongoMemoryServer.create();
    await mongoose.connect(mongoServer.getUri());
  });

  afterEach(async function disconnectFromMongoDB() {
    await mongoose.disconnect();
    await mongoose.connection.close();
  });

  // =====================================
  // Signup
  // =====================================

  describe("signup", () => {
    it("with valid data it shoud create new user", async () => {
      let response = await supertest(app).post("/api/v2/auth/signup").send({
        username: "james",
        email: "james@gmail.com",
        password: "testingTEST123@",
      });

      expect(response.status).toBe(201);
      expect(response.body).toMatchObject({
        user: { email: "james@gmail.com" },
      });

      expect(response.body.user.passwordDigest).toBeUndefined();
      expect(response.body).toMatchSnapshot({
        accessToken: expect.any(String),
        user: {
          ...response.body.user,
          _id: expect.any(String),
          userId: expect.any(String),
          createdAt: expect.any(String),
          updatedAt: expect.any(String),
        },
      });

      expect(
        response.headers["set-cookie"].find((cookie: string) =>
          cookie.includes("refreshToken")
        )
      ).toBeDefined();
    });

    describe("user already exists", () => {
      beforeEach(async () => {
        await User.create({
          username: "james",
          email: "james@gmail.com",
          passwordDigest: "testingTEST@123",
        });
      });

      it("if user already then given an user exists error", async () => {
        let response = await supertest(app).post("/api/v2/auth/signup").send({
          username: "james",
          email: "james@gmail.com",
          password: "testingTEST@123",
        });

        expect(response.status).toEqual(400);
        expect(response.body).toMatchSnapshot();
      });
    });
  });

  // =====================================
  // Login
  // =====================================

  describe("login", () => {
    describe("when user doesn't exists", () => {
      it("then give a 404 response", async () => {
        let response = await supertest(app)
          .post("/api/v2/auth/login")
          .send({ email: "james@gmail.com", password: "testingTEST@123" });

        expect(response.status).toEqual(404);
        expect(response.body).toMatchSnapshot();
      });
    });

    describe("when user exists", () => {
      beforeEach(async () => {
        await User.create({
          email: "james@gmail.com",
          passwordDigest: "testingTEST@123",
        });
      });

      it("then successfully login", async () => {
        let response = await supertest(app)
          .post("/api/v2/auth/login")
          .send({ email: "james@gmail.com", password: "testingTEST@123" });

        expect(response.status).toEqual(200);

        expect(response.body).toMatchSnapshot({
          accessToken: expect.any(String),
          user: {
            ...response.body.user,
            _id: expect.any(String),
            userId: expect.any(String),
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
          },
        });

        expect(
          response.headers["set-cookie"].find((cookie: string) =>
            cookie.includes("refreshToken")
          )
        ).toBeDefined();
      });
    });
  });

  describe("access-token", () => {
    let refreshToken: string;

    beforeEach(async () => {
      let user = await User.create({
        email: "james@gmail.com",
        passwordDigest: "testingTEST@123",
      });

      refreshToken = user.getRefreshToken();
    });

    describe("when refresh token is valid", () => {
      it("then successfully get a new access token", async () => {
        let response = await supertest(app)
          .get("/api/v2/auth/access-token")
          .set("Cookie", [`refreshToken=${refreshToken}`]);

        expect(response.status).toEqual(200);

        expect(response.body).toMatchSnapshot({
          accessToken: expect.any(String),
          user: {
            ...response.body.user,
            _id: expect.any(String),
            userId: expect.any(String),
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
          },
        });
      });
    });
  });
});
