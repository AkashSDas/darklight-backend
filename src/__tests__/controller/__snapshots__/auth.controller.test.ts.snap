// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth controller access-token when refresh token is valid then successfully get a new access token 1`] = `
{
  "accessToken": Any<String>,
  "user": {
    "__v": 0,
    "_id": Any<String>,
    "active": false,
    "banned": false,
    "createdAt": Any<String>,
    "email": "james@gmail.com",
    "enrolledCourses": [],
    "oauthProviders": [],
    "roles": [
      "base",
    ],
    "updatedAt": Any<String>,
    "userId": Any<String>,
    "verified": false,
  },
}
`;

exports[`auth controller login when user doesn't exists then give a 404 response 1`] = `
{
  "message": "User not found",
}
`;

exports[`auth controller login when user exists then successfully login 1`] = `
{
  "accessToken": Any<String>,
  "user": {
    "__v": 0,
    "_id": Any<String>,
    "active": false,
    "banned": false,
    "createdAt": Any<String>,
    "email": "james@gmail.com",
    "enrolledCourses": [],
    "oauthProviders": [],
    "roles": [
      "base",
    ],
    "updatedAt": Any<String>,
    "userId": Any<String>,
    "verified": false,
  },
}
`;

exports[`auth controller signup user already exists if user already then given an user exists error 1`] = `
{
  "message": "User already exists",
}
`;

exports[`auth controller signup with valid data it shoud create new user 1`] = `
{
  "accessToken": Any<String>,
  "user": {
    "__v": 0,
    "_id": Any<String>,
    "active": false,
    "banned": false,
    "createdAt": Any<String>,
    "email": "james@gmail.com",
    "enrolledCourses": [],
    "oauthProviders": [],
    "roles": [
      "base",
    ],
    "updatedAt": Any<String>,
    "userId": Any<String>,
    "username": "james",
    "verified": false,
  },
}
`;
